packages/starlight/user-components/Tabs.astro

```diff
@@ -9,8 +9,67 @@ interface Props {
 const { syncKey } = Astro.props;
 const panelHtml = await Astro.slots.render('default');
 const { html, panels } = processPanels(panelHtml);
+
+/**
+ * Synced tabs are persisted across page using `localStorage`. The script used to restore the
+ * active tab for a given sync key has a few requirements:
+ *
+ * - The script should only be included when at least one set of synced tabs is present on the page.
+ * - The script should be inlined to avoid a flash of invalid active tab.
+ * - The script should only be included once per page.
+ *
+ * To do so, we keep track of whether the script has been rendered using a variable stored using
+ * `Astro.locals` which will be reset for each new page. The value is tracked using an untyped
+ * symbol on purpose to avoid Starlight users to get autocomplete for it and avoid potential
+ * clashes with user-defined variables.
+ *
+ * The restore script defines a custom element `starlight-tabs-restore` that will be included in
+ * each set of synced tabs to restore the active tab based on the persisted value using the
+ * `connectedCallback` lifecycle method. To ensure this callback can access all tabs and panels for
+ * the current set of tabs, the script should be rendered before the tabs themselves.
+ */
+const isSynced = syncKey !== undefined;
+const didRenderSyncedTabsRestoreScriptSymbol = Symbol.for('starlight:did-render-synced-tabs-restore-script');
+// @ts-expect-error - See above
+const shouldRenderSyncedTabsRestoreScript = isSynced && Astro.locals[didRenderSyncedTabsRestoreScriptSymbol] !== true;
+
+if (isSynced) {
+	// @ts-expect-error - See above
+	Astro.locals[didRenderSyncedTabsRestoreScriptSymbol] = true
+}
 ---
 
+{/* Inlined to avoid a flash of invalid active tab. */}
+{shouldRenderSyncedTabsRestoreScript && <script is:inline>
+	(() => {
```

Comment:
Nit: whitespace will be preserved inside this inline script tag once rendered, so could maybe indent by one less tab?

(Also fantasising a bit about a home-made system to minify inline scripts — imagine we had an `inline-scripts/` directory that got loaded, minified, and available as a string each from a virtual module, e.g. `import { tabsRestore } from 'virtual:starlight-inline-scripts';`. Then these scripts could look like `<script is:inline set:html={tabsRestore} />` or something.)

Reply:
Quite the interesting idea, I would guess a bit easier to setup when we have a build system, but definitely interesting 🧠

Reply:
Actually, scratch that, I just quickly tried locally using the JS API of esbuild and a Vite plugin and it works, even supports watch mode. Altho, this makes me wonder, could Astro auto minify inline scripts (I guess there are some conditions to this) 🤔

Reply:
Yeah, I’ve been chatting about that a bit with the team. Posted a proposal here: https://github.com/withastro/roadmap/discussions/995